doc/compiler/aot/SymbolValidationManager.md

```diff
@@ -165,6 +165,46 @@ load, these symbols don't need to be validation records stored in
 the SCC. Therefore, the SVM, in its constructor, initializes the maps
 with these IDs and the associated symbols.
 
+### Well-known Classes
+
+Some system classes are loaded very early and are referred to by most
+JIT compilations. The SVM has a list of such classes, termed "well-known
+classes." It will store a single copy of their class chain offsets in
+the SCC. For each AOT-compiled method using the SVM, the emitted
+relocation data needs just one offset into the SCC in order to refer to
+the shared data. This not only deduplicates the offsets; it also allows
+many otherwise necessary records to be eliminated.
+
+In particular, well-known classes never need SystemClassByName,
+ClassByName, or ClassFromCP records. Because these latter two have
+beholders, eliminating them is a bit more difficult than eliminating
+SystemClassByName. In order to make them unnecessary, the SVM checks for
+every class it encounters that the class can see every well-known class,
```

Comment:
>the SVM checks for every class it encounters that the class can see every well-known class

This doesn't result in potential load failures (outside of the class not being loaded yet) because the names of these well known classes are only allowed to be loaded by the bootstrap classloader right?

Reply:
They'll only be *defined* by the bootstrap loader, but in order for code belonging to other loaders to name these classes, those loaders have to be able to find them (by delegating).

Since we're not allowed to run Java code, it might be possible to fail to find a class even from a loader that *will* delegate and get it at runtime, if that loader hasn't previously been asked about that class. If so, we could fail loads as a result.

However, the entire "well-known classes" technique carries an increased risk of load failure, because it creates dependencies on classes that the compilation hasn't naturally queried for.